src/ApiService/ApiService/onefuzzlib/Metrics.cs

```diff
@@ -0,0 +1,53 @@
+﻿using System.Text.Json;
+using System.Text.Json.Serialization;
+using Microsoft.OneFuzz.Service.OneFuzzLib.Orm;
+
+namespace Microsoft.OneFuzz.Service {
+
+    public record CustomMetric(
+         string name,
+         int value,
+         Dictionary<string, string> customDimensions
+    );
+
+
+    public interface IMetrics {
+        void SendMetric(int metricValue, BaseEvent customDimensions);
+
+        void LogMetric(BaseEvent metric);
+    }
+
+    public class Metrics : IMetrics {
+        private readonly ILogTracer _log;
+        private readonly IOnefuzzContext _context;
+        private readonly JsonSerializerOptions _options;
+
+        public Metrics(ILogTracer log, IOnefuzzContext context) {
+            _context = context;
+            _log = log;
+            _options = new JsonSerializerOptions(EntityConverter.GetJsonSerializerOptions()) {
+                DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
+            };
+            _options.Converters.Add(new RemoveUserInfo());
+        }
+
+        public void SendMetric(int metricValue, BaseEvent customDimensions) {
+            var metricType = customDimensions.GetEventType();
+
+            _ = _options.PropertyNamingPolicy ?? throw new ArgumentException("Serializer _options not available.");
+
+            var metricTypeSnakeCase = _options.PropertyNamingPolicy.ConvertName($"{metricType}");
+
+            var dimensionString = JsonSerializer.Serialize(customDimensions, customDimensions.GetType(), _options);
+            var dimensionDict = JsonSerializer.Deserialize<Dictionary<string, string>>(dimensionString);
+
+            _log.Metric($"{metricTypeSnakeCase}", metricValue, dimensionDict);
+            LogMetric(customDimensions);
+        }
```

Comment:
try this instead. It should save you the round trip of serialize deserialize
```C#
var dimensionDict = JsonSerializer.SerializeToNode(customDimensions, customDimensions.GetType(), _options ).AsObject().ToDictionary(kvp => kvp.Key, kvp =>  kvp.Value.ToString());
```

Reply:
Oh wow this is incredibly helpful. Thank you Cheick.